public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/class-rest-feedback-controller.php

```diff
@@ -232,4 +236,86 @@ public function create_item_permissions_check( $request ) {
 
 		return true;
 	}
+
+	/**
+	 * Additional preparing of feedback data before processing and saving it.
+	 *
+	 * @param array $prepared_feedback
+	 * @param array $request
+	 *
+	 * @return array
+	 */
+	public function preprocess_comment( $prepared_feedback, $request ) {
+		if ( isset( $request['meta'] ) ) {
+			$validated_meta = validate_feedback_meta( $request['meta'] );
+
+			if ( is_wp_error( $validated_meta ) ) {
+				return $validated_meta;
+			}
+
+			$prepared_feedback['comment_meta'] = $validated_meta;
+		}
+
+		// These indexes might be missing, but need to be present for `wp_allow_comment`.
+		$prepared_feedback = wp_parse_args(
+			$prepared_feedback,
+			array(
+				'comment_author_url' => '',
+				'comment_agent'      => '',
+				'comment_type'       => COMMENT_TYPE, // We set this later, but it still needs to be here.
+			)
+		);
+
+		return $prepared_feedback;
+	}
+
+	/**
+	 * Incorporate meta values into the duplicate check.
+	 *
+	 * Note that this assumes the meta data in `$comment_data` has already been validated by `validate_feedback_meta`.
+	 *
+	 * @param string|int|null $duplicate_id Unused.
+	 * @param array           $comment_data The current comment data that might be a duplication.
+	 *
+	 * @return bool
+	 */
+	public function duplicate_check( $duplicate_id, $comment_data ) {
+		$args = array(
+			'number'     => 1,
+			'type'       => COMMENT_TYPE,
+			'status'     => array( 'approve', 'hold', 'spam' ),
+			'post_id'    => $comment_data['comment_post_ID'],
+			'meta_query' => array( 'relation' => 'AND' ),
+		);
+
+		if ( isset( $comment_data['user_id'] ) ) {
+			$args['user_id'] = $comment_data['user_id'];
+		} else {
+			$args['author_email'] = $comment_data['comment_author_email'];
+		}
+
+		$schema = get_feedback_meta_field_schema();
+
+		foreach ( $comment_data['comment_meta'] as $key => $value ) {
+			$args['meta_query'][] = array(
+				'key'   => $key,
+				'value' => $value,
+				// This might need to change if we introduce a meta field with a type other than string or integer.
+				'type'  => ( 'string' === $schema[ $key ]['type'] ) ? 'CHAR' : 'NUMERIC',
```

Comment:
We could add a note in the schema function that only string & integer are supported? The only other case I could think of would be boolean, and that could easily be saved as string `true` or int `1`.

Reply:
Good idea. Added a comment: 8fcb7df5